### ISSUE TITLE
why the continuous write to async streaming crash?

### ISSUE THREAD

**Issue by fxrrrr** [grpc/grpc#4007]
In my option, when i want to send stream msg to client, i can just do like this:

ServerAsyncWriter< TReply > stream_;

for(int i=0;i<msg.size();i++){
  stream_.Write(msg[i],this);
}

but when send more than 5 'Write' just one time, it result in a crash here:
//call.c
static grpc_cq_completion *allocate_completion(grpc_call *call) {
  gpr_uint8 i;
  gpr_mu_lock(&call->completion_mu);
  for (i = 0; i < GPR_ARRAY_SIZE(call->completions); i++) {
    if (call->allocated_completions & (1u << i)) {
      continue;
    }
    /* NB: the following integer arithmetic operation needs to be in its
     * expanded form due to the "integral promotion" performed (see section
     * 3.2.1.1 of the C89 draft standard). A cast to the smaller container type
     * is then required to avoid the compiler warning */
    call->allocated_completions =
        (gpr_uint8)(call->allocated_completions | (1u << i));
    gpr_mu_unlock(&call->completion_mu);
    return &call->completions[i];
  }
  GPR_UNREACHABLE_CODE(return NULL);   //<<<<<------------------- CRASH HERE
  return NULL;
}

it seems like due to this limit:
//call.c
/** The maximum number of completions possible.
    Based upon the maximum number of individually queueable ops in the batch
   api:
      - initial metadata send
      - message send
      - status/close send (depending on client/server)
      - initial metadata recv
      - message recv
      - status/close recv (depending on client/server) */
#define MAX_CONCURRENT_COMPLETIONS 6  //<<<<<---------------------

so, what's wrong with my usage? need i do some other operation after each 'Write'?
try to trigger 'done_completion' by call 'AsyncNext'?

**Comment by fxrrrr** [comment]
the crash stack:

GasRunner.exe!abort() 行 88	C
GasRunner.exe!allocate_completion(grpc_call * call) 行 439	C
GasRunner.exe!finish_batch(grpc_exec_ctx * exec_ctx, grpc_call * call, int success, void * tag) 行 1635	C
GasRunner.exe!unlock(grpc_exec_ctx * exec_ctx, grpc_call * call) 行 714	C
GasRunner.exe!call_on_done_send(grpc_exec_ctx * exec_ctx, void * pc, int success) 行 894	C
GasRunner.exe!grpc_exec_ctx_flush(grpc_exec_ctx * exec_ctx) 行 50	C
GasRunner.exe!grpc_exec_ctx_finish(grpc_exec_ctx * exec_ctx) 行 60	C
GasRunner.exe!grpc_call_start_batch(grpc_call * call, const grpc_op * ops, unsigned int nops, void * tag, void * reserved) 行 1901	C
GasRunner.exe!grpc::Server::PerformOpsOnCall(grpc::CallOpSetInterface * ops, grpc::Call * call) 行 426	C++
GasRunner.exe!grpc::Call::PerformOps(grpc::CallOpSetInterface * ops) 行 89	C++
GasRunner.exe!grpc::ServerAsyncWriter<AsyncFlow::FCRunningStatus>::Write(const AsyncFlow::FCRunningStatus & msg, void * tag) 行 377	C++

**Comment by fxrrrr** [comment]
I think the ' grpc_cq_completion completions[MAX_CONCURRENT_COMPLETIONS];'
was designed to store some thing like the 'tag'.
Each call to 'SendInitialMetadata' , 'Write' , 'Finish' ,etc will give the user a chance to set a 'tag'.
and the storage info will be used later as to do some callback.
In sync mode, each 'Write' will deal with the storage info immediately, so can reset them between each 'Write'.
but in async mode, if the 'Write' operation was used more than once and with different 'tag'.
It can not just store them all with a 'MAX_CONCURRENT_COMPLETIONS' size of space.
For most case, the 'tag' passed to 'Write' will be the same?

**Comment by ctiller** [comment]
So for async writes we have a (poorly documented admittedly) contract that
says you're only allowed to have one outstanding at a time.
This means that you need to wait for the completion to the first write
before starting the next one - grpc doesn't do any internal buffering.

**Comment by fxrrrr** [comment]
So what should i do exactly after each write? Call 'cq_.AsyncNext'?  I find that before the crash, the message was sended()

**Comment by fxrrrr** [comment]
... find that before the crash, the message was sended(in tcp_window.c) and the client recv all the writed messages.

**Comment by hsaliak** [comment]
@ctiller is there any documentation updates need to make?

**Comment by joostmeijles** [comment]
I have a gRPC Server instance that serves multiple streaming clients using a shared completion queue.
When messages get written to the clients (using `grpc::ServerAsyncWriter<T>::Writer` call) shortly after each other I get an assertion error:
`E0307 11:54:41.724590260    5917 server.cc:442]              assertion failed: GRPC_CALL_OK == result`

The result variable contains the following value:
`/** there is already an outstanding read/write operation on the call */
  GRPC_CALL_ERROR_TOO_MANY_OPERATIONS,`

Is this the same problem as described in this issue?
If so, this seems to be a show-stopper to me, as I imagine that synchronously writing messages to an asynchronous stream seems to be basic required functionality?

What would be the workaround? For example, how do I determine that there is an outstanding message to send?

**Comment by joostmeijles** [comment]
I didn't get any response, will this be fixed for the 2016/04/01 milestone?

If not, can you (@vjpai ?) please provide or hint a workaround? This issue is really blocking me.

**Comment by a-teammate** [comment]
@ctiller why was this closed?

Im asking bc i dont understand whats the use case for an async call which always needs a blocking procedure directly afterwards..

Only usage i can imagine is when a lot of other blocking work needs to be done on your current thread and networking is really low priority ( `for(;;) { Write(); /*do blocking work*/ cq->Next();}`).

Otherwise I'd assume one usually wants to do the blocking work (completionqueue->Next()) in a background thread, and call the async->Write() in the current one (which needs to be nonblocking).

However with only one Write() allowed this would mean we would need some kind of manual syncing again between those threads.
So what will people do?  
I'd guess usually youd grab some kind of (lockless or threadsafe) queue (maybe even FIFO if thats important for the use case) and just put writedata into it in your important thread and in a background thread you enque it and feed the rpc connection.  And it would look like `async->Write(Data); completionque->Next();` in the background thread, but then theres no point in using the async call, or is it?

(And since theres already a queue inside grpc, why dont hide it and allow multiple Write() calls?)

**Comment by ctiller** [comment]
So it was closed because there's no bug here, and I'd prefer not to have
issues open for things that aren't.

It probably warrants some design notes though, so here goes:

We allow only one write at a time to ensure there's an avenue for providing
pushback against applications. Not doing so (by introducing a queue)
defeats flow control and we have good data that says that that will lead to
applications OOMing themselves - and having no way available to fix it.

There are two patterns for using the async API that seem to be relatively
common.

The first is having a set of threads running against one or more completion
queues in a server. All server work is performed in these threads: when a
request comes in, any work it needs to do is performed asynchronously, with
a state machine driving which calls to make. Here, the tags passed to the
completion queue are frequently some closure object (std::function<void()>*
for instance) that records how to progress the state machine when the
operation completes.

The second is when fanning out a request to multiple back-ends in a
synchronous API server. In this case an incoming request might trigger many
sub-requests before it can be completed. Each sub-request is made
asynchronously, and the overall request thread blocks until all completions
have been drained from the completion queue created to service *only that*
request.
